productivity-suite/limitations-and-improvements.md

```diff
@@ -0,0 +1,63 @@
+
+This method requires the fragments’ authors to be in control and tweak the build application in order for its type module scripts to follow the required convention, this not only can be difficult and different for different technologies but can even sometimes be unfeasible in situations where complex and non configurable bundling solutions are in place.
+
+## Fragments’ base paths
+
+Our fragments are implemented and deployed as standalone applications, and as any other web application when they need to access assets that is done based on the base path of their url.
+
+For example, if a web application is deployed at `https://www.example.com` and requires the `images/logo.png` asset (e.g. it includes such an HTML tag: `<img src=”images/logo.png” >`) the browser will try to fetch such asset from `https://www.example.com/images/logo.png`.
+
+This is of course the correct behavior, but it is problematic for our strategy since when we use fragment content in the application we create a misalignment between the base paths (in our example above, the gateway worker would be the one receiving the `images/logo.png` request instead of the fragment worker itself and it wouldn’t know how to handle it).
+
+We’ve thought of different solutions for this issues, which include:
+
+- Always providing assets via external absolute paths (e.g. using CDNs)
+- Modifying the HTML, JS and CSS imports at the fragment build time
+- Using appropriate conventions so that the gateway worker can discern which asset belongs to which fragment and forwarding the requests to the proper workers
+
+We haven’t had the need to use such assets in our demo’s fragments, so we haven’t yet specifically decided on a solution to this problem.
+
+We will however need to address and decide on the best approach for this issue soon as it impacts not only this specific strategy but micro frontend fragment based applications in general.
+
+## Current frameworks’ assumptions
+
+During our experimentations we discovered that most modern frameworks, although compatible with fragments, can make different assumptions about the state and scope of ownership of the HTML document.
+
+They can assume that all the rest of the document is populated solely by fragments implemented by the framework, that it is valid to manipulate the document itself and add variables to the global scope or that they will always be present in the page (not being dynamically added and removed).
```

Comment:
In particular we had problems with cleanly unloading fragments from the page, and also re-initializing fragments when they are loaded dynamically.

Reply:
I've slightly rewarded your sentence and added it to the file, please have a look 🙂 (https://github.com/cloudflare/workers-web-experiments/pull/58/commits/0901aa5133ad213ab77e980e5fcd2e5804b9c3d0)